service/controller/v17/encrypter/kms/kms.go

```diff
@@ -51,24 +51,53 @@ func (e *Encrypter) EnsureCreatedEncryptionKey(ctx context.Context, customObject
 		return microerror.Mask(err)
 	}
 
+	var keyAlias string
+	{
+		clusterID := key.ClusterID(customObject)
+		keyAlias = toAlias(clusterID)
+	}
+
+	var oldKeyScheduledForDeletion bool
 	{
 		e.logger.LogCtx(ctx, "level", "debug", "message", "finding out encryption key")
 
 		_, err := e.describeKey(ctx, customObject)
 		if IsKeyNotFound(err) {
 			e.logger.LogCtx(ctx, "level", "debug", "message", "did not find encryption key")
 
+		} else if IsKeyScheduledForDeletion(err) {
+			e.logger.LogCtx(ctx, "level", "debug", "message", "found encryption key")
+			e.logger.LogCtx(ctx, "level", "debug", "message", "current encryption key is scheduled for deletion and will be recreated")
+			oldKeyScheduledForDeletion = true
+
 		} else if err != nil {
 			return microerror.Mask(err)
 
 		} else {
-
 			e.logger.LogCtx(ctx, "level", "debug", "message", "found encryption key")
 			e.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
 			return nil
 		}
 	}
 
+	if oldKeyScheduledForDeletion {
+		// In such case we just delete the alias so we can alias newly
+		// created key.
+
+		e.logger.LogCtx(ctx, "level", "debug", "message", "deleting old encryption key alias")
+
+		in := &kms.DeleteAliasInput{
+			AliasName: aws.String(keyAlias),
+		}
+
+		_, err = ctlCtx.AWSClient.KMS.DeleteAlias(in)
```

Comment:
Lets say the deletion fails and we retry the resource. Do we still get to here? I think this might be handled differently.

Reply:
I'm not sure if I get you right. After retry if the alias is still pointing to the key which is scheduled for deletion we should delete the alias. When alias is deleted we can't find any key by this alias we won't get into this `if` statement. Does that make sense or you meant something else? We can hang on that.

Reply:
When this ensures the alias deletion on retries this is fine. I am just suspicious when there is state in memory and conditionals and errors and how does the system behave when all falls apart.

Reply:
That's why I wrote e2e test for that.

Reply:
The e2e tests do not inject failures purposefully to ensure what I described.

Reply:
I just think such a failure is likely to happen during cluster recreation.